src/util.c

```diff
@@ -552,6 +552,28 @@ int string2d(const char *s, size_t slen, double *dp) {
     return 1;
 }
 
+int doubleIsLongLong(double d, long long *out) {
+#if (DBL_MANT_DIG >= 52) && (LLONG_MAX == 0x7fffffffffffffffLL)
+        /* Check if the float is in a safe range to be casted into a
+         * long long. We are assuming that long long is 64 bit here.
+         * Also we are assuming that there are no implementations around where
+         * double has precision < 52 bit.
+         *
+         * Under this assumptions we test if a double is inside an interval
+         * where casting to long long is safe. Then using two castings we
+         * make sure the decimal part is zero. If all this is true we use
+         * integer printing function that is much faster. */
+        double min = -4503599627370495; /* (2^52)-1 */
+        double max = 4503599627370496; /* -(2^52) */
```

Comment:
the comments are wrong, but i'm also uncertain the values are correct.
in integer's 2's complement, the negative side is the larger one (unlike this code).
however, doubles have a distinct sign bit, so i think their range to the positive and negative is the same.
i think the numbers should be `4503599627370495` and `-4503599627370495`

but i also think this check doesn't necessarily have to be here, and can't cause much harm.
i.e. i suppose we can also safely keep a value of 4503599627370495*4 as long without precision loss (but not *3).
and also, even if the range check would have been too big (allowing some values that can't be stored in long), the later conversion check is what actually decides, so this range check is just an optimization to avoid casting overheads?

Reply:
see my last commit and the updated top comment.

Reply:
The same changes need to be made for https://github.com/redis/redis/blob/b8eb2a73408fa3b8845760857dd6fcccb62107fe/src/rdb.c#L601:L602